fps/src/main/scala/chapter03/exercise3_16.scala

```diff
@@ -0,0 +1,19 @@
+
+package chapter03
+
+import chapter03.listImplementation.*
+import chapter03.listImplementation.List.*
+
+import scala.annotation.tailrec
+
+
+object exercise3_16 extends App:
+
+  def append[A](as: List[A], a: A): List[A] =
+    reverse(Cons(a, reverse(as)))
+
+
+  def addOne(as: List[Int]): List[Int] =
+    foldLeft(as,List.empty[Int])((a,i) => append(a, i+1))
```

Comment:
now that we gave `append` a fitting type signature, ie. `(List[A],Int) => List[Int]` it "fits" the `foldLeft` folding function value `f` and as a result can also be written `foldLeft(as,List.empty[Int])(append)`..

now isn't that neat ;))

Reply:
I added the (a,i) => append(a,i+1) because I didn't know how to do I + 1 otherwise. Is it possible to still do this then?

Reply:
yes - if the parameter order of a method fit the tuple order of a function's input type you can omit calling the method explicitly but instead just pass the method as a parameter, eg.

```
scala> def plus(l: Int, r: Int): Int = l + r
def plus(l: Int, r: Int): Int
                                                                                                                            
scala> List(1,2,3,4,5).foldLeft(0)(plus)
val res2: Int = 15
```

instead of:

```
scala> List(1,2,3,4,5).foldLeft(0)((l,r) => plus(l,r))
val res2: Int = 15
```